File: leetcode/python/844_backspace_string_compare.py
from typing import cast


class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        s1: list[str] = []
        s2: list[str] = []
        for c in s:
            if c != '#':
                s1.append(c)
            elif len(s1) > 0:
                s1.pop()
        for c in t:
            if c != '#':
                s2.append(c)
            elif len(s2) > 0:
                s2.pop()
            
        return ''.join(s1) == ''.join(s2)
            
if __name__ == '__main__':
    s = Solution()
    print(s.backspaceCompare("ab#c", "ad#c"))
    print(s.backspaceCompare("ab##", "c#d#"))
    print(s.backspaceCompare("a#c", "b"))
    print(s.backspaceCompare("y#fo##f", "y#f#o##f"))
    

File: leetcode/rust/src/844_backspace_string_compare.rs
impl Solution {
    pub fn backspace_compare(s: String, t: String) -> bool {
        let s = s.chars().collect::<Vec<_>>();
        let t = t.chars().collect::<Vec<_>>();

        let mut processed_s = vec![];
        for c in s {
            if c == '#' {
                processed_s.pop();
            } else {
                processed_s.push(c);
            }
        }
        
        let mut processed_t = vec![];
        for c in t {
            if c == '#' {
                processed_t.pop();
            } else {
                processed_t.push(c);
            }
        }

        processed_s == processed_t
    }
}